primitives/src/unique_vec.rs

```diff
@@ -0,0 +1,334 @@
+//! Module with [`UniqueVec`] type and related functional.
+
+#[cfg(not(feature = "std"))]
+use alloc::{borrow::ToOwned as _, format, string::String, vec::Vec};
+use core::{
+    borrow::Borrow,
+    fmt::{Debug, Display},
+    ops::Deref,
+};
+
+use iroha_schema::IntoSchema;
+use parity_scale_codec::{Decode, Encode};
+use serde::{Deserialize, Serialize};
+
+/// Creates a [`UniqueVec`](unique_vec::UniqueVec) from a list of values.
+///
+/// Works like [`vec!`] macro, but does not accept syntax for repeated values
+/// and might return [`Result`].
+#[macro_export]
+macro_rules! unique_vec {
+    () => {
+        $crate::unique_vec::UniqueVec::new()
+    };
+    ($($x:expr),+ $(,)?) => {
+        $crate::unique_vec::UniqueVec::try_from(vec![$($x),+])
+    };
+}
+
+/// Duplicated value at index {index}
+#[cfg_attr(feature = "std", derive(thiserror::Error))]
+#[derive(Debug, displaydoc::Display, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
+pub struct DuplicationError {
+    index: usize,
+}
+
+/// Wrapper type for [`Vec`] which ensures that all elements are unique.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Encode, IntoSchema)]
+pub struct UniqueVec<T: PartialEq>(Vec<T>);
+
+impl<T: PartialEq> Default for UniqueVec<T> {
+    fn default() -> Self {
+        Self(Vec::default())
+    }
+}
+
+impl<T: PartialEq> UniqueVec<T> {
+    /// Create new [`UniqueVec`].
+    #[must_use]
+    pub const fn new() -> Self {
+        Self(Vec::new())
+    }
+
+    /// Try construct [`UniqueVec`] from iterator.
+    ///
+    /// # Errors
+    ///
+    /// - If there are duplicates in the sequence.
+    pub fn try_from_iter<I>(iter: I) -> Result<Self, DuplicationError>
+    where
+        I: IntoIterator<Item = T>,
+    {
+        let mut v = UniqueVec::new();
+        v.try_extend(iter)?;
+        Ok(v)
+    }
+
+    /// Try extend [`UniqueVec`] from iterator.
+    /// If an error occurs, all previously added elements will remain in the [`UniqueVec`].
+    ///
+    /// # Errors
+    ///
+    /// - If there are duplicates in the sequence.
+    pub fn try_extend<I>(&mut self, iter: I) -> Result<(), DuplicationError>
```

Comment:
I would prefer if you just implemented `Extend` trait instead of providing `try_extend` method. It should be fine to remove duplicates in the process of extending. Maps or sets do a similar thing where they reject some of the `other` collection members silently

Reply:
Hm, haven't thought about that. Makes sense, will do

Reply:
Wait. We want it to fail to deserialize if there are any duplicates. Why don't we want the same explicit error in other cases?

Reply:
Just checked how serde deserialization works for `HashSet`. It ignores all duplicates. Here is a workaround:
https://github.com/serde-rs/serde/issues/1607

Reply:
Should I provide the same kind of API?

Reply:
so, our choice here is to either:
1. have all custom fallible methods
2. always silently remove duplicates
 - in this case we can deserialize list which contains duplicates, and implement `FromIterator` and `Extend`
 
 > Should I provide the same kind of API?

I would prefer it that way. To me 2. would make more sense for `UniqueVec` if we look at it separately from it's use as a vec of peers. For a particular case where throwing error is required it can be done via a custom deserializing function
 
Did you consider using [itertools::all_unique](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.all_unique)

Reply:
> Did you consider using

I did not, will take a look, thanks

> To me 2 would make more sense for `UniqueVec`

Ok, that's ok for me. I just wanted to be sure that these two options are clear for us and we should just choose one of them. Will do.